packages/flutter_tools/lib/src/build_system/tools/asset_transformer.dart

```diff
@@ -136,6 +143,70 @@ final class AssetTransformer {
   }
 }
 
+
+// A wrapper around [AssetTransformer] to support hot reload of transformed assets.
+final class DevelopmentAssetTransformer {
+  DevelopmentAssetTransformer({
+    required FileSystem fileSystem,
+    required AssetTransformer transformer,
+    required Logger logger,
+  })  : _fileSystem = fileSystem,
+        _transformer = transformer,
+        _logger = logger;
+
+  final AssetTransformer _transformer;
+  final FileSystem _fileSystem;
+  final Pool _transformationPool = Pool(4);
+  final Logger _logger;
+
+  /// Re-transforms an asset and returns a [DevFSContent] that should be synced
+  /// to the attached device in its place.
+  ///
+  /// Returns `null` if any of the transformation subprocesses failed.
+  Future<DevFSContent?> retransformAsset({
+    required String inputAssetKey,
+    required DevFSContent inputAssetContent,
+    required List<AssetTransformerEntry> transformerEntries,
+    required String workingDirectory,
+  }) async {
+    final File output = _fileSystem.systemTempDirectory.childFile('retransformerInput-$inputAssetKey');
+    ErrorHandlingFileSystem.deleteIfExists(output);
+    late File inputFile;
+    bool cleanupInput = false;
+    Uint8List result;
+    PoolResource? resource;
+    try {
+      resource = await _transformationPool.request();
+      if (inputAssetContent is DevFSFileContent) {
+        inputFile = inputAssetContent.file as File;
+      } else {
+        inputFile = _fileSystem.systemTempDirectory.childFile('retransformerInput-$inputAssetKey');
+        inputFile.writeAsBytesSync(await inputAssetContent.contentsAsBytes());
+        cleanupInput = true;
+      }
+      final AssetTransformationFailure? failure = await _transformer.transformAsset(
```

Comment:
We end up doing a redundant copy of the initial input file here, but it's unclear if it's worth having a boolean flag parameter or a separate method on `AssetTransformer` that transforms the asset without doing the defensive copy. Alternatively, the code calling `AssetTransformer::transformAsset` can do the defensive copy themselves, but that would be even more copy-paste code in all the places that write assets.

Reply:
Yeah, I'm leaning towards keeping the code simpler. this LGTM.